unit/analyses/call_graph.cpp

```diff
@@ -0,0 +1,113 @@
+
+#include <iostream>
+
+#include <catch.hpp>
+
+#include <analyses/call_graph.h>
+
+#include <util/symbol_table.h>
+#include <util/std_code.h>
+
+#include <goto-programs/goto_convert_functions.h>
+
+SCENARIO("call_graph",
+  "[core][util][call_graph]")
+{
+
+  symbol_tablet symbol_table;
+
+  code_typet void_function_type;
+
+  {
+    code_blockt calls;
+    code_function_callt call1;
+    call1.function()=symbol_exprt("A", void_function_type);
+    code_function_callt call2;
+    call2.function()=symbol_exprt("B", void_function_type);
+    calls.move_to_operands(call1);
+    calls.move_to_operands(call2);
+
+    symbolt a_function;
+    a_function.name="A";
+    a_function.type=void_function_type;
+    a_function.mode=ID_java;
+    a_function.value=calls;
+    symbol_table.add(a_function);
+  }
+
+  {
+    code_blockt calls;
+    code_function_callt call1;
+    call1.function()=symbol_exprt("C", void_function_type);
+    code_function_callt call2;
+    call2.function()=symbol_exprt("D", void_function_type);
+    calls.move_to_operands(call1);
+    calls.move_to_operands(call2);
+
+    symbolt b_function;
+    b_function.name="B";
+    b_function.type=void_function_type;
+    b_function.mode=ID_java;
+    b_function.value=calls;
+    symbol_table.add(b_function);
+  }
+
+  {
+    symbolt c_function;
+    c_function.name="C";
+    c_function.type=void_function_type;
+    c_function.mode=ID_java;
+    c_function.value=code_skipt();
+    symbol_table.add(c_function);
+  }
+
+  {
+    symbolt d_function;
+    d_function.name="D";
+    d_function.type=void_function_type;
+    d_function.mode=ID_java;
+    d_function.value=code_skipt();
+    symbol_table.add(d_function);
+  }
+
+  call_grapht call_graph_from_symbol_table(symbol_table);
+  call_grapht inverse_call_graph_from_symbol_table=
+    call_graph_from_symbol_table.get_inverted();
+
+
+  GIVEN("A call graph constructed from a symbol table")
+  {
+    WHEN("Checking the call graph")
+    {
+      REQUIRE(call_graph_from_symbol_table.graph.size()==4);
+      REQUIRE(call_graph_from_symbol_table.graph.count("A")==2);
+      REQUIRE(call_graph_from_symbol_table.graph.count("B")==2);
```

Comment:
I think this could probably do with a few more checks (e.g. that `graph["A"] = { A, B }`, `graph.count("C") == 0`)